lib/teleterm/gateway/gateway_test.go

```diff
@@ -45,3 +49,42 @@ func TestCLICommandUsesCLICommandProvider(t *testing.T) {
 
 	require.Equal(t, "foo/bar", command)
 }
+
+func TestGatewayStart(t *testing.T) {
+	hs := httptest.NewTLSServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {}))
+	t.Cleanup(func() {
+		hs.Close()
+	})
+
+	gateway, err := New(
+		Config{
+			TargetName:   "foo",
+			TargetURI:    uri.NewClusterURI("bar").AppendDB("foo").String(),
+			TargetUser:   "alice",
+			Protocol:     defaults.ProtocolPostgres,
+			CertPath:     "../../../fixtures/certs/proxy1.pem",
+			KeyPath:      "../../../fixtures/certs/proxy1-key.pem",
+			Insecure:     true,
+			WebProxyAddr: hs.Listener.Addr().String(),
+		},
+		mockCLICommandProvider{},
+	)
+	require.NoError(t, err)
+	require.Equal(t, gateway, gateway)
+
+	wait := make(chan error)
+
+	go func() {
+		err := gateway.Open()
+		wait <- err
+	}()
+
+	defer func() {
+		// Make sure Open() is called.
+		time.Sleep(time.Millisecond * 200)
+
+		err := gateway.Close()
+		require.NoError(t, err)
+		require.NoError(t, <-wait)
+	}()
+}
```

Comment:
I already talked with Zac about this pattern under https://github.com/gravitational/teleport/pull/14074#issuecomment-1175229671. `time.Sleep` is bad but I don't know what would be a better approach with this API. Maybe I should just reword this comment so that it doesn't suggest that sleeping actually makes sure that `Open()` is called.

Reply:
sidenote: `(*Gateway).Open` and `(*LocalProxy).Start` should probably both be called `Serve` or something along those lines, IMO

Reply:
second sidenote: why not just `net.Dial` the address to confirm that the other side has actually accepted the connection?

Reply:
@espadolini Oh that's a good idea, would something like what I pushed in 1efd5ef work? The APIs in the net package are one level lower than what I'm used to.

> sidenote: `(*Gateway).Open` and `(*LocalProxy).Start` should probably both be called `Serve` or something along those lines, IMO

I don't really want to rewrite ALPN proxy code right now but I can certainly take care of `Gateway.Open`.